File: src/os/os.cpp
#include "../coroutine.h"
#include "os.h"
#include "process.h"

namespace flame {
namespace os {
	void declare(php::extension_entry& ext) {
		ext
			.function<interfaces>("flame\\os\\interfaces")
			.function<spawn>("flame\\os\\spawn", {
				{"executable", php::TYPE::STRING},
				{"arguments", php::TYPE::ARRAY, false, true},
				{"options", php::TYPE::ARRAY, false, true},
			})
			.function<exec>("flame\\os\\exec", {
				{"executable", php::TYPE::STRING},
				{"arguments", php::TYPE::ARRAY, false, true},
				{"options", php::TYPE::ARRAY, false, true},
			});
		process::declare(ext);
	}
	php::value interfaces(php::parameters& params) {
		struct ifaddrs* addr;
		if(getifaddrs(&addr) != 0) {
			throw php::exception(zend_ce_error, std::strerror(errno));
		}else if(!addr) {
			return php::array(0);
		}
		php::array data(4);
		// 用 shared_ptr 作自动释放保证
		std::shared_ptr<struct ifaddrs> autofree(addr, freeifaddrs);
		do {
			if(addr->ifa_addr->sa_family != AF_INET && addr->ifa_addr->sa_family != AF_INET6) continue;
			php::string name(addr->ifa_name);
			php::array  info(2);
			info.set("family", addr->ifa_addr->sa_family == AF_INET ? "IPv4" : "IPv6");
			char address[NI_MAXHOST];
			if(getnameinfo(addr->ifa_addr,
				addr->ifa_addr->sa_family == AF_INET ? sizeof(struct sockaddr_in) : sizeof(struct sockaddr_in6),
				address, NI_MAXHOST, nullptr, 0, NI_NUMERICHOST) != 0) {
				throw php::exception(zend_ce_error, gai_strerror(errno));
			}
			info.set("address", address);
			if(data.exists(name)) {
				php::array iface = data.get(name);
				iface.set(iface.size(), info);
			}else{
				php::array iface(2);
				iface.set(iface.size(), info);
				data.set(name, iface);
			}
		}while((addr = addr->ifa_next) != nullptr);

		return std::move(data);
	}
	php::value spawn(php::parameters& params) {
		php::object proc(php::class_entry<process>::entry());
		process* proc_ = static_cast<process*>(php::native(proc));
		proc_->exit_ = false;

		auto env = boost::this_process::environment();
		std::string exec = params[0].to_string();
		if(exec[0] != '.' && exec[0] != '/') {
			exec = boost::process::search_path(exec).native();
		}
		std::vector<std::string> argv;
		std::string cwdv = boost::filesystem::current_path().native();
		if(params.size() > 1) {
			php::array args = params[1];
			for(auto i=args.begin(); i!=args.end(); ++i) {
				argv.push_back( i->second.to_string() );
			}
		}
		if(params.size() > 2) {
			php::array opts = params[2];
			php::array envs = opts.get("env");
			if(envs.typeof(php::TYPE::ARRAY)) {
				for(auto i=envs.begin(); i!=envs.end(); ++i) {
					env[i->first.to_string()] = i->second.to_string();
				}
			}
			php::string cwds = opts.get("cwd");
			if(cwds.typeof(php::TYPE::STRING)) {
				cwdv = cwds.to_string();
			}
		}
		
		boost::process::child c(exec, boost::process::args = argv, env, context,
			boost::process::start_dir = cwdv,
			boost::process::std_out > proc_->out_,
			boost::process::std_err > proc_->err_,
			boost::process::on_exit = [proc, proc_] (int exit_code, const std::error_code&) {
			proc_->exit_ = true;
			if(proc_->co_wait) {
				proc_->co_wait->resume();
			}
		});
		proc.set("pid", c.id());
		proc_->c_ = std::move(c);
		return proc;
	}
	php::value exec(php::parameters& params) {
		php::object proc = spawn(params);
		process* proc_ = static_cast<process*>(php::native(proc));
		proc_->co_wait = coroutine::current;
		proc_->co_wait->stack(php::value([proc, proc_] (php::parameters& params) -> php::value {
			return proc_->out_.get();
		}));
		return coroutine::async();
	}
}
}

File: doc/os.php
<?php
/**
 * 提供操作系统相关信息读取、进程操作等功能封装
 */
namespace flame\os;
/**
 * 信号
 */
const SIGTERM = 15;
/**
 * 信号
 */
const SIGKILL = 9;
/**
 * 信号
 */
const SIGINT  = 2;
/**
 * 信号
 */
const SIGUSR1 = 10;
/**
 * 信号
 */
const SIGUSR2 = 12;

/**
 * 获取当前系统网卡绑定地址等信息
 * @example
 *  array(2) {
 *      ["lo"]=>
 *      array(1) {
 *          [0]=>
 *          array(2) {
 *          ["family"]=>
 *          string(4) "IPv4"
 *          ["address"]=>
 *          string(9) "127.0.0.1"
 *          }
 *      }
 *      ["eth0"]=>
 *      array(1) {
 *          [0]=>
 *          array(2) {
 *          ["family"]=>
 *          string(4) "IPv4"
 *          ["address"]=>
 *          string(13) "10.110.16.197"
 *          }
 *      }
 *      }
 */
function interfaces():array {}
/**
 * 异步启动进程
 * @return 进程对象
 */
function spawn(string $command, array $argv = [], array $options = []):process {}
/**
 * 调用上述 spawn() 异步启动进程, 并等待其结束, 返回进程标准输出
 * @param array $options 目前可用的选项如下：
 *  * "cwd" - string - 工作路径;
 *  * "env" - array - 环境变量，K/V 结构文本;
 * @return string 进程标准输出内容
 */
function exec(string $command, array $argv, array $options = []):string {}

/**
 * 进程对象
 */
class process {
    /**
     * 向进程发送指定信号
     */
    function kill(int $signal = SIGTERM) {}
    /**
     * 等待进程结束
     */
    function wait() {}
    /**
     * 获取进程标准输出（若进程还未结束需要等待）
     */
    function stdout():string {}
    /**
     * 获取进程错误输出（若进程还未结束需要等待）
     */
    function stderr():string {}
}